src/esp32/pwm.rs

```diff
@@ -0,0 +1,209 @@
+    pub fn get_duty_pct(&self) -> f64 {
+        let max_duty = self.ledc_driver.get_max_duty();
+        (self.ledc_driver.get_duty() as f64) / (max_duty as f64)
+    }
+
+    pub fn get_frequency(&self) -> u32 {
+        unsafe { ledc_get_freq(SpeedMode::LowSpeed.into(), self.ledc_driver.timer()) }
+    }
+
+    pub fn set_frequency(&mut self, frequency_hz: u32) -> anyhow::Result<()> {
+        let mut ledc_manager = LEDC_MANAGER.lock().unwrap();
+        ledc_manager.drop_timer(self.timer_number, self.channel);
+        let timer_number = ledc_manager.take_timer(frequency_hz, self.channel)?;
+        let timer: ledc_timer_t = (timer_number as u8).into();
+        match esp_idf_sys::esp!(unsafe {
+            ledc_set_freq(SpeedMode::LowSpeed.into(), timer, frequency_hz)
+        }) {
+            Ok(val) => { Ok(val) },
+            Err(err) => {
+                ledc_manager.drop_timer(timer_number, self.channel);
+                Err(err)
+            }
+        }?;
+        self.timer_number = timer_number;
+        Ok(())
+    }
+}
+
+struct LedcManager {
+    channel_pin_subscriptions: HashMap<i32, i32>,
+    timer_frequencies: [Option<u32>; 4],
+    timer_channels: [Vec<PwmChannel>; 4]
+}
+
+impl LedcManager {
+    fn new() -> Self {
+        Self {
+            channel_pin_subscriptions: HashMap::new(),
+            timer_frequencies: [None, None, None, None],
+            timer_channels: [Vec::new(), Vec::new(), Vec::new(), Vec::new()]
+        }
+    }
+
+    fn find_timer_by_frequency(&mut self, frequency_hz: u32) -> Option<usize> {
+        self.timer_frequencies.iter().position(|&x| x == Some(frequency_hz))
+    }
+
+    fn take_timer(&mut self, frequency_hz: u32, channel: PwmChannel) -> anyhow::Result<usize> {
+        let timer_number = if self.timer_frequencies.iter().all(|&x| x.is_some()) {
+            // in the future we might want to instead re-purpose the least recently used timer?
+            let err_string = format!(
+                "no more than 4 different PWM frequencies simultaneously allowed, available frequencies: {}, {}, {}, {}",
+                self.timer_frequencies[0].unwrap(),
+                self.timer_frequencies[1].unwrap(),
+                self.timer_frequencies[2].unwrap(),
+                self.timer_frequencies[3].unwrap(),
+            );
+            self.find_timer_by_frequency(frequency_hz).context(err_string)?
+        } else {
+            let timer_number = self
+                .timer_frequencies
+                .iter()
+                .position(|&x| match x {
+                    Some(x) => x == frequency_hz,
+                    None => false
+                });
+            match timer_number {
+                Some(timer_number) => {
+                    self.timer_channels[timer_number].push(channel);
+                    timer_number
+                },
+                None => {
+                    let timer_number = self.timer_channels.iter().position(|x| x.is_empty()).unwrap();
```

Comment:
can we return an error instead of unwrap?

Reply:
due to the condition in L. 36, it's impossible that this is an error. Would it be acceptable to make it a match statement and use unreachable?